File: shared/src/utils/slug.ts
/** Gera slug URL-safe a partir de um texto (remove acentos, espaços -> hifen). */
export function slugify(input: string): string {
  return input
    .normalize('NFD')
    .replace(/[̀-ͯ]/g, '') // remove diacriticos combinantes
    .toLowerCase()
    .trim()
    .replace(/[^a-z0-9\s-]/g, '')
    .replace(/\s+/g, '-')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '');
}

/** SKU interno deterministico e legivel: TP-<CATEGORIA>-<SEQ/rand>. */
export function buildInternalSku(category: string | undefined, seed: string): string {
  const cat = (category ?? 'GEN')
    .slice(0, 3)
    .toUpperCase()
    .replace(/[^A-Z]/g, 'X');
  const suffix = seed
    .replace(/[^a-zA-Z0-9]/g, '')
    .slice(-6)
    .toUpperCase()
    .padStart(6, '0');
  return `TP-${cat}-${suffix}`;
}

File: backend/app/shared/utilities/slug.py
import re
import unicodedata


def slugify(input_: str) -> str:
    normalized = unicodedata.normalize("NFD", input_)
    without_accents = "".join(char for char in normalized if unicodedata.category(char) != "Mn")
    lowered = without_accents.lower().strip()
    safe = re.sub(r"[^a-z0-9\s-]", "", lowered)
    dashed = re.sub(r"\s+", "-", safe)
    collapsed = re.sub(r"-+", "-", dashed)
    return collapsed.strip("-")


def build_internal_sku(category: str | None, seed: str) -> str:
    raw_category = (category or "GEN")[:3].upper()
    cat = re.sub(r"[^A-Z]", "X", raw_category)
    suffix = re.sub(r"[^a-zA-Z0-9]", "", seed)[-6:].upper().rjust(6, "0")
    return f"TP-{cat}-{suffix}"
